File: GraphInDepthPractice/dfs.cpp
#include <bits/stdc++.h>

using namespace std;

class graph
{
    int num_vertex;
    int time;
    vector<vector<int>> adjList;

    //this is the recursive approach for DFS
    void DFS_rec(int s, vector<bool> &visited, vector<int> &pre, vector<int> &post);
    void DFS_iter(int s, vector<bool> &visited);

public:
    graph(int v) : num_vertex(v), adjList(v), time(0)
    {
    }
    void add_edge(int u, int v)
    {
        adjList[u].push_back(v);
    }

    void DFS(int s);
};

void graph::DFS(int s)
{

    //Doing BFS in Recursive Way
    vector<bool> visited(num_vertex, false);
    cout << "Recursive Way: " << endl;
    vector<int> pre(num_vertex);
    vector<int> post(num_vertex);
    for (int i = 0; i < num_vertex; i++)
    {
        if (!visited[i])
        {
            DFS_rec(i, visited, pre, post);
        }
    }

    //Doing DFS in Iterative Way
    vector<bool> visited2(num_vertex, false);
    cout << "Iterative Way:" << endl;
    for (int i = 0; i < num_vertex; i++)
    {
        if (!visited2[i])
        {
            DFS_iter(i, visited2);
        }
    }

    //After DFS, it is printing the Pre and Post Time for Each and Every Node
    cout<<"After DFS, Pre and Post Time of Each Node:"<<endl;
    for(int i=0; i<num_vertex; i++)
    {
        cout<<i<<" "<<"Pre:"<<pre[i]<<" Post:"<<post[i]<<endl; 
    }
}
void graph::DFS_rec(int s, vector<bool> &visited, vector<int> &pre, vector<int> &post)
{
    visited[s] = true;
    cout << s << endl;
    //pre time or starting time for a node
    pre[s] = time++;
    for (auto u : adjList[s])
    {
        if (!visited[u])
        {
            DFS_rec(u, visited, pre, post);
        }
    }
    //post time or ending time of a node
    post[s] = time++;
}
void graph::DFS_iter(int s, vector<bool> &visited)
{
    stack<int> stk;
    stk.push(s);
    visited[s] = true;

    while (!stk.empty())
    {
        int u = stk.top();
        cout << u << " ";
        stk.pop();
        for (auto i : adjList[u])
        {
            if (!visited[i])
            {
                visited[i] = true;
                stk.push(i);
            }
        }
    }
}

int main()
{
    graph g(5);

    g.add_edge(0, 1);
    g.add_edge(0, 3);
    g.add_edge(1, 2);
    g.add_edge(3, 2);
    g.add_edge(3, 4);

    g.DFS(0);
}

File: GraphInDepthPractice/dfs.py
class graph:
    def __init__(self, v):
        self.numVertex = v
        self.adjList = [[]for i in range(v)]

    def addEdge(self, u, v):
        self.adjList[u].append(v)
    def DFS_rec(self, s, visited):
        visited[s] = True
        print(s)
        for i in self.adjList[s]:
            if not visited[i]:
                self.DFS_rec(i, visited)
    def DFS_iter(self, s, visited):
        stk = []
        stk.append(s)
        visited[s] = True
        while len(stk)>0:
            u = stk[-1]
            stk.pop()
            print(u)
            for i in self.adjList[u]:
                if not visited[i]:
                    stk.append(i)
                    visited[i] = True

    def DFS(self, s):
        print("Recursive DFS: ")
        visited = [False]*self.numVertex
        for i in range(self.numVertex):
            if not visited[i]:
                self.DFS_rec(i, visited)
        print("Iterative DFS")
        visited2 = [False]*self.numVertex
        for j in range(self.numVertex):
            if not visited2[j]:
                self.DFS_iter(j, visited2)


g = graph(5)
g.addEdge(0, 1)
g.addEdge(0, 3)
g.addEdge(1, 2)
g.addEdge(3, 2)
g.addEdge(3, 4)

g.DFS(0)